modules/core/src/main/java/org/apache/ignite/spi/indexing/IndexingSpi.java

```diff
@@ -84,7 +91,56 @@
      * @param expirationTime Expiration time or 0 if never expires.
      * @throws IgniteSpiException If failed.
      */
-    public void store(@Nullable String cacheName, Object key, Object val, long expirationTime) throws IgniteSpiException;
+    public default void store(@Nullable String cacheName, Object key, Object val, long expirationTime) throws IgniteSpiException {
+        // No-op.
+    }
+
+    /**
+     * Updates index with new row. Note that key is unique for cache, so if cache contains multiple indexes
+     * the key should be removed from indexes other than one being updated.
+     *
+     * @param cctx Cache context.
+     * @param newRow cache row to store in index.
+     * @param prevRow optional cache row that will be replaced with new row.
+     */
+    public void store(GridCacheContext<?, ?> cctx, CacheDataRow newRow, @Nullable CacheDataRow prevRow,
```

Comment:
I don't sure it's a good idea to request an internal classes from public API

Reply:
Replace IgniteIndexing (IndexSpi) with IndexProcessor